Test for errors during large message send/receipt + retry on send fail

joshuef opened the pull request:
Close issue https://github.com/maidsafe/safe_network/issues/69

- adds retry attempts on new connections
- adds retry atetmpts on sending messages
- adds test for sending many 1mb messages

joshuef committed:
tests: add checks for errors during large message flow

joshuef committed:
chore: send result on disconnect

lionel-faber committed:
tests: fix test

joshuef committed:
feat: retry on connect, retry on send

joshuef commented on src/connection_deduplicator.rs:
```diff
@@ -73,15 +73,21 @@ impl ConnectionDeduplicator {
+            Error::Connection(source) => match source {
+                quinn::ConnectionError::ConnectionClosed(_) => Self::QuinnConnectionClosed,
+                err => {
+                    tracing::warn!("** ???? ?the conn err: {:?}", err);
```
sorry, this shouldn't have been marked as ready for review yet!

Yoga07 commented on src/endpoint.rs:
```diff
@@ -41,7 +41,10 @@ const PORT_FORWARD_TIMEOUT: u64 = 30;
+const STANDARD_CHANNEL_SIZE: usize = 10000;
+
+/// Max number of attempts for connection retries
+const MAX_ATTEMPTS: usize = 10;
```
Would `10` be too much? Meaning, it can take some time to complete all 10 retries if there's the receiver is gone surely. Maybe a smaller number like 3 or 5?

maqi commented on src/connections.rs:
```diff
@@ -41,9 +52,11 @@ impl Connection {
+        self.handle_error(send_msg(&mut send_stream, msg.clone()).await)
             .await?;
 
+        trace!("Message sent");
```
this log will not be that useful for production code. 
maybe remove it or make it log some more useful info?

joshuef replied:
removed

Yoga07 commented on src/endpoint.rs:
```diff
@@ -500,7 +517,18 @@ impl Endpoint {
+        while attempts < MAX_ATTEMPTS && res.is_err() {
+            trace!("send attempt # {:?}", attempts);
+            attempts += 1;
+            sleep(Duration::from_millis(500)).await;
```
Could we have this configurable? `RETRY_INTERVAL`?

joshuef commented on src/connections.rs:
```diff
@@ -41,9 +52,11 @@ impl Connection {
     pub async fn send_uni(&self, msg: Bytes) -> Result<()> {
         let mut send_stream = self.handle_error(self.quic_conn.open_uni().await).await?;
-        self.handle_error(send_msg(&mut send_stream, msg).await)
+        self.handle_error(send_msg(&mut send_stream, msg.clone()).await)
```
Bytes clone is cheap, so not that much of an issue. But aye looks like its no longer needed here

joshuef committed:
chore: small tweaks

Return quinn error from deduplicator
remove an unneeded clone
reduce max attempts

joshuef committed:
feat: Add config option for retry interval

joshuef commented on .gitignore:
```diff
@@ -3,3 +3,4 @@
 Cargo.lock
 *.*~
 .idea
+*.log
```
how do you mean?

maqi replied:
I mean it needs an extra line break at the end of file.
otherwise it will show this red icon.

joshuef committed:
chore: remove log

maqi commented on src/endpoint.rs:
```diff
@@ -500,7 +517,18 @@ impl Endpoint {
+        let mut attempts: usize = 0;
+        let mut res = self.try_send_message(msg.clone(), dest).await;
+
+        while attempts < MAX_ATTEMPTS && res.is_err() {
```
we will try resends for 10 times, then on disconnection, try re-connect for 10 times as well ?
and each takes an interval of 0.5s, which makes a total of 50s to detect a really failed connection ?

joshuef replied:
Could be. (worth nothing, this has already been updated to `MAX_ATTEMPTS=5` ). 

What's an acceptable time here? How many tries do we want? 

50s doesn't seem too bad. _to be sure_, vs just a 30s timeout eg

maqi replied:
mainly just confirming this setup is ok.

joshuef commented on src/connections.rs:
```diff
@@ -193,15 +201,22 @@ pub(super) fn listen_for_incoming_messages(
     let _ = tokio::spawn(async move {
-        debug!("qp2p another incoming listerner spawned");
-        let _ = future::join(
+        match future::try_join(
```
Yeh the other future will be cancelled as I understand.

joshuef commented on src/connections.rs:
```diff
@@ -288,19 +311,24 @@ async fn read_on_bi_streams(
+                            "Failed reading from a bi-stream for peer {:?} with: {:?}",
                             peer_addr, err
                         );
                         break;
```
As I can figure, it's a question of "can this connection still be used". If so, we do not error.

joshuef replied:
So we were breaking before here, so that's not a change. 

As I understand, here we've received the bytes and it's the processing that fails (some message fails), but the connection is still valid, so we don't want to err out and cause a disconnect.

joshuef committed:
ci: no cargo verbose and ignore reachability for coverage check

joshuef committed:
ci: skip the large messages test for coverage check

lionel1704 commented on src/connections.rs:
```diff
@@ -28,6 +28,17 @@ pub(crate) struct Connection {
     remover: ConnectionRemover,
 }
 
+/// Disconnection events, and the result that led to disconnection.
```
Outdated comment?
